# KNN Classifier atau Klasifikasi

**Nama  : Shohibun Naufal**

**NIM   : 21041110013**

**Kelas : Penambangan Data-C**

KNN (K-Nearest Neighbors) adalah salah satu metode machine learning dalam klasifikasi dan regresi. Metode ini memprediksi kelas atau nilai target dari sebuah data baru dengan membandingkannya dengan data-data terdekat di sekitarnya dalam ruang fitur.

Secara singkat, KNN bekerja dengan cara mencari K tetangga terdekat dari data baru, di mana K adalah suatu bilangan bulat yang ditentukan sebelumnya. Kemudian, KNN akan mengambil mayoritas kelas dari K tetangga terdekat tersebut sebagai prediksi kelas untuk data baru (untuk klasifikasi) atau menghitung rata-rata nilai target dari K tetangga terdekat tersebut sebagai prediksi nilai target untuk data baru (untuk regresi).



1. Mengimport library yang diperlukan

In [4]:
import pandas as pd
import numpy as np
from sklearn.neighbors import KNeighborsRegressor 
from sklearn.preprocessing import LabelEncoder

from sklearn.impute import KNNImputer
from sklearn.preprocessing import OrdinalEncoder


2. Import data kita

In [5]:
df = pd.read_csv('https://gist.githubusercontent.com/shohibunnaufal/16420efa3ecb0d483c7274e9c867dcb9/raw/4fe0bc2e59dbe0a5bb8265ba4a2b22545b58a9b4/heartdisease.csv')

jumlah_baris = df.shape[0]
df.head(jumlah_baris)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

3. Mengecek nilai yang hilang/kosong pada data

In [6]:
df_copy = df.copy()
df_copy.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

4. Ganti nilai yang mengandung tanda tanya (?) menjadi nilai kosong

In [7]:
df_copy = df_copy.replace('?',np.nan)

5. Cek lagi berapa nilai yang kosong pada data

In [8]:
df_copy.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

6. Setelah itu kita lihat type datanya, jika type datanya bukan numerik/float maka perlu diganti type datanya

In [9]:
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


7. Melakukan Normalisasi data mengubah setiap nilai ke dalam rentang 0 hingga 1,dengan metode min-max dan juga metode simple feature scalling(z-score) .namun untuk Kolom target(kolom species ) tak akan di normalisasi

In [11]:
target_column = df_copy['target']
df_copy = df_copy[['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']].copy()
df_copy = (df_copy - df_copy.min())/(df_copy.max() - df_copy.min())
df_copy['target'] = target_column.copy()
df_copy.head(jumlah_baris)

age  sex        cp  trestbps      chol  fbs  restecg   thalach  \
0    0.708333  1.0  1.000000  0.481132  0.244292  1.0      0.0  0.603053   
1    0.166667  1.0  0.666667  0.339623  0.283105  0.0      0.5  0.885496   
2    0.250000  0.0  0.333333  0.339623  0.178082  0.0      0.0  0.770992   
3    0.562500  1.0  0.333333  0.245283  0.251142  0.0      0.5  0.816794   
4    0.583333  0.0  0.000000  0.245283  0.520548  0.0      0.5  0.702290   
..        ...  ...       ...       ...       ...  ...      ...       ...   
298  0.583333  0.0  0.000000  0.433962  0.262557  0.0      0.5  0.396947   
299  0.333333  1.0  1.000000  0.150943  0.315068  0.0      0.5  0.465649   
300  0.812500  1.0  0.000000  0.471698  0.152968  1.0      0.5  0.534351   
301  0.583333  1.0  0.000000  0.339623  0.011416  0.0      0.5  0.335878   
302  0.583333  0.0  0.333333  0.339623  0.251142  0.0      0.0  0.786260   

     exang   oldpeak  slope    ca      thal  target  
0      0.0  0.370968    0.0  0.00  0.333333       1  
1      0.0  0.564516    0.0  0.00  0.666667       1  
2      0.0  0.225806    1.0  0.00  0.666667       1  
3      0.0  0.129032    1.0  0.00  0.666667       1  
4      1.0  0.096774    1.0  0.00  0.666667       1  
..     ...       ...    ...   ...       ...     ...  
298    1.0  0.032258    0.5  0.00  1.000000       0  
299    0.0  0.193548    0.5  0.00  1.000000       0  
300    0.0  0.548387    0.5  0.50  1.000000       0  
301    1.0  0.193548    0.5  0.25  1.000000       0  
302    0.0  0.000000    0.5  0.25  0.666667       0  

[303 rows x 14 columns]

8.Pada langkah-langkah terakhir dalam inisialisasi objek knn dari kelas KNeighborsRegressor, kita menggunakan algoritma K-Nearest Neighbors untuk melakukan regresi. Untuk memastikan hasil yang konsisten, kita melakukan randomisasi indeks dataset menggunakan np.random.permutation() dengan seed 1.

Kemudian, kita membagi dataset menjadi data training sebesar 75% dan data testing sebesar 25% menggunakan variabel train_index. Data training kemudian disimpan di variabel training_data, sementara data testing disimpan di variabel testing_data.

Setelah itu, kita melakukan training model menggunakan data training dengan memanggil method fit() pada objek knn, dengan argumen input berupa fitur-fitur yang digunakan (dalam hal ini sepal_length, sepal_width, petal_length, dan petal_width) dan label yaitu species.

Setelah model terlatih, kita melakukan prediksi pada data testing dengan memanggil method predict() pada objek knn dengan argumen input berupa fitur-fitur dari data testing yang ingin diprediksi.

Kemudian, untuk mengukur akurasi model, kita menghitung persentase data testing yang berhasil diprediksi dengan menggunakan threshold default sebesar 0.5.

Terakhir, hasil akurasi model dicetak untuk dianalisis lebih lanjut

In [13]:
knn = KNeighborsRegressor()

# before we split, make sure you randomize the dataset
np.random.seed(1)
index_shuffled = np.random.permutation(df_copy.index)
shuffled_car_data = df_copy.loc[index_shuffled]

# We will use 75% of the data for training and the remaining 25% for testing
train_index = int(shuffled_car_data.shape[0] * 0.75)
training_data = shuffled_car_data.iloc[0:train_index]
testing_data = shuffled_car_data.iloc[train_index:]

# fit the model with the training data
knn.fit(training_data[['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']], training_data['target'])

# predict the target of the testing data
prediction = knn.predict(testing_data[['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']])

# calculate the accuracy of the model
accuracy = np.mean(np.abs(prediction - testing_data['target']) <= 0.5)

print(f"Akurasi model KNN: {accuracy:.2%}")


Akurasi model KNN: 92.11%


## Kesimpulan

Menurut hasil diatas, akurasi data yang diperoleh dengan model KNN adalah 92.11%